crates/red_knot_python_semantic/src/types.rs

```diff
@@ -306,3 +312,359 @@ pub struct IntersectionType<'db> {
+        }
+    }
+
+    #[test]
+    fn build_union() {
+        let db = setup_db();
+        let t0 = Type::IntLiteral(0);
+        let t1 = Type::IntLiteral(1);
+        let Type::Union(union) = UnionTypeBuilder::new(&db).add(t0).add(t1).build() else {
+            panic!("expected a union");
+        };
+
+        assert_eq!(union.elements_vec(&db), &[t0, t1]);
+    }
+
+    #[test]
+    fn build_union_single() {
+        let db = setup_db();
+        let t0 = Type::IntLiteral(0);
+        let ty = UnionTypeBuilder::new(&db).add(t0).build();
+
+        assert_eq!(ty, t0);
+    }
+
+    #[test]
+    fn build_union_empty() {
+        let db = setup_db();
+        let ty = UnionTypeBuilder::new(&db).build();
+
+        assert_eq!(ty, Type::Never);
+    }
+
+    #[test]
+    fn build_union_never() {
+        let db = setup_db();
+        let t0 = Type::IntLiteral(0);
+        let ty = UnionTypeBuilder::new(&db).add(t0).add(Type::Never).build();
+
+        assert_eq!(ty, t0);
+    }
+
+    #[test]
+    fn build_union_flatten() {
+        let db = setup_db();
+        let t0 = Type::IntLiteral(0);
+        let t1 = Type::IntLiteral(1);
+        let t2 = Type::IntLiteral(2);
+        let u1 = UnionTypeBuilder::new(&db).add(t0).add(t1).build();
+        let Type::Union(union) = UnionTypeBuilder::new(&db).add(u1).add(t2).build() else {
+            panic!("expected a union");
+        };
+
+        assert_eq!(union.elements_vec(&db), &[t0, t1, t2]);
+    }
+
+    impl<'db> IntersectionType<'db> {
+        fn pos_vec(self, db: &'db TestDb) -> Vec<Type<'db>> {
+            self.positive(db).into_iter().collect()
+        }
+
+        fn neg_vec(self, db: &'db TestDb) -> Vec<Type<'db>> {
+            self.negative(db).into_iter().collect()
+        }
+    }
+
+    #[test]
+    fn build_intersection() {
+        let db = setup_db();
+        let t0 = Type::IntLiteral(0);
+        let t1 = Type::IntLiteral(1);
+        let Type::Intersection(inter) = IntersectionTypeBuilder::new(&db)
+            .add_positive(t0)
+            .add_negative(t1)
+            .build()
+        else {
+            panic!("expected to be an intersection");
+        };
+
+        assert_eq!(inter.pos_vec(&db), &[t0]);
+        assert_eq!(inter.neg_vec(&db), &[t1]);
+    }
+
+    #[test]
+    fn build_intersection_flatten_positive() {
+        let db = setup_db();
+        let t0 = Type::IntLiteral(0);
+        let t1 = Type::IntLiteral(1);
+        let t2 = Type::IntLiteral(2);
+        let i0 = IntersectionTypeBuilder::new(&db)
+            .add_positive(t0)
+            .add_negative(t1)
+            .build();
+        let Type::Intersection(inter) = IntersectionTypeBuilder::new(&db)
+            .add_positive(t2)
+            .add_positive(i0)
+            .build()
+        else {
+            panic!("expected to be an intersection");
+        };
+
+        assert_eq!(inter.pos_vec(&db), &[t2, t0]);
+        assert_eq!(inter.neg_vec(&db), &[t1]);
```

Comment:
You might want to add a comment here about how we could simplify this further once we have greater understanding of assignability/subtyping; otherwise it might be confusing if somebody else tries to implement it later and this test starts failing.

...Hmm, but perhaps that applies for quite a few tests being added here. Not sure.

Reply:
Ah hm, yeah, good point -- my favorite option would be to rework this to use types where it is correct not to simplify; that might actually not be too hard, using `Any`. I'll make that change, thanks for pointing this out!

Reply:
I also noticed while updating the tests that the self-canceling code should not apply to `Any` or `Unknown`: `Any & ~Any` does not simplify to `Never`, because the two `Any` are both unknown types and might not be the same type; it should instead simplify to `Any`. I just added a TODO for this, it doesn't need to be fixed in this PR.